src/passes/analysis/verilog.cpp

```diff
@@ -1,93 +1,589 @@
+        this->verilator_debug = true;
+    }
 }
 
 std::string Passes::Verilog::ID = "verilog";
-bool Passes::Verilog::runOnInstanceGraphNode(InstanceGraphNode& node) {
-  
-  //Create a new Vmodule for this node
-  Module* m = node.getModule();
-  
-  vmods.addModule(m);
-  return false;
+
+// Helper function that prepends a prefix contained in json metadata if it
+// exists
+std::string make_name(std::string name, json metadata) {
+    if (metadata.count("prefix") > 0) {
+        name = metadata["prefix"].get<std::string>() + name;
+    }
+    return name;
 }
 
-void Passes::Verilog::writeToStream(std::ostream& os) {
-  for (auto module : vmods.vmods) {
-    if (vmods._inline && module->inlineable) {
-      continue;
+// Converts a CoreIR `Value` type into a Verilog literal
+std::unique_ptr<vAST::Expression> convert_value(Value *value) {
+    if (auto arg_value = dyn_cast<Arg>(value)) {
+        return std::make_unique<vAST::Identifier>(arg_value->getField());
+    } else if (auto int_value = dyn_cast<ConstInt>(value)) {
+        return std::make_unique<vAST::NumericLiteral>(int_value->toString());
+    } else if (auto bool_value = dyn_cast<ConstBool>(value)) {
+        return std::make_unique<vAST::NumericLiteral>(
+            std::to_string(uint(bool_value->get())));
+    } else if (auto bit_vector_value = dyn_cast<ConstBitVector>(value)) {
+        BitVector bit_vector = bit_vector_value->get();
+        return std::make_unique<vAST::NumericLiteral>(
+            bit_vector.hex_digits(), bit_vector.bitLength(), false, vAST::HEX);
+    } else if (auto string_value = dyn_cast<ConstString>(value)) {
+        return std::make_unique<vAST::String>(string_value->toString());
     }
-    WriteModuleToStream(module, os);
-  }
+    coreir_unreachable();
 }
 
-void Passes::Verilog::writeToFiles(const std::string& dir,
-                                   std::unique_ptr<std::string> product_file) {
-  std::vector<std::string> products;
-  for (auto module : vmods.vmods) {
-    if (vmods._inline && module->inlineable) continue;
-    const std::string filename = module->modname + ".v";
-    products.push_back(filename);
-    const std::string full_filename = dir + "/" + filename;
-    std::ofstream fout(full_filename);
-    ASSERT(fout.is_open(), "Cannot open file: " + full_filename);
-    WriteModuleToStream(module, fout);
-    fout.close();
-  }
-  // Write out the product list, if requested.
-  if (!product_file) return;
-  std::ofstream fout(*product_file);
-  ASSERT(fout.is_open(), "Cannot open file: " + *product_file);
-  for (const auto& product : products) {
-    fout << product << "\n";
-  }
-  fout.close();
+// Given a signal named `id` and a type `type`, wrap the signal name in a
+// `Vector` node if the signal is of type Array
+std::variant<std::unique_ptr<vAST::Identifier>, std::unique_ptr<vAST::Vector>>
+process_decl(std::unique_ptr<vAST::Identifier> id, Type *type) {
+    if (type->getKind() == Type::TK_Array) {
```

Comment:
You should use `if (ArrayType* array_type = dyn_cast<ArrayType>(type)) {`

In general getKind() should not be used much. dyn_cast<> and isa<> should be used instead.
See the doc/ProgrammingGuide.md